### PR TITLE
Merkle Memberships Program; Clean up 

### PR DESCRIPTION
**Pull request by chfanghr**

### CODE REVIEW

File: src/library/plugin/pluginType.ts [hunk 1]
```diff
@@ -4,20 +4,20 @@ import z from 'zod';
 
 // Interfaces used on the server side.
 
-export interface IMinAuthPlugin<PublicInputsArgs, Output> {
+export interface IMinAuthPlugin<FetchPublicInput, Output> {
```
Comment (adamczykm) [inline src/library/plugin/pluginType.ts]: Just a rename

File: src/library/tools/pluginServer/config.ts [hunk 2]
```diff
@@ -1,10 +1,10 @@
-import { IMinAuthPlugin, IMinAuthPluginFactory } from "plugin/pluginType";
+import { IMinAuthPlugin, IMinAuthPluginFactory } from '../../../library/plugin/pluginType';
 import z from "zod";
 import env from 'env-var';
 import fs from 'fs';
 import yaml from 'yaml';
-import { SimplePreimagePlugin } from "plugins/simplePreimage/plugin";
-import { SimplePasswordTreePlugin } from "plugins/passwordTree/plugin";
+import { SimplePreimagePlugin } from "./plugins/simplePreimage/server";
+import { MemberSetPlugin } from "./plugins/passwordTree/server";
```
Comment (adamczykm) [inline src/library/tools/pluginServer/config.ts]: Just a rename

File: src/library/tools/pluginServer/index.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import express, { Request, Response } from 'express';
 import bodyParser from 'body-parser';
 import { JsonProof, verify } from 'o1js';
-import { IMinAuthPlugin } from 'plugin/pluginType';
+import { IMinAuthPlugin } from '../../../library/plugin/pluginType';
```
Comment (adamczykm) [inline src/library/tools/pluginServer/index.ts]: The previous path has not been working for me.

File: src/plugins/passwordTree/common/passwordTreeProgram.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+export class MerkleRoot extends Struct({
+    root: Field
+}) {};
+
+
+export class ProvePasswordInTreeOutput extends Struct({
+    recursiveMekleRootHash: Field,
+}) {};
+
+// Prove knowledge of a preimage of a hash in a merkle tree.
```
Comment (adamczykm) [inline src/plugins/passwordTree/common/passwordTreeProgram.ts]: The zk program has changed. The other one was not doing any meaningful ZK operation. This one hides the user. 
The rest of the code needs fixing now.

**Review by adamczykm — CHANGES_REQUESTED** [review]

**Comment by adamczykm** [comment]
Additionally, I think that we will need to rethink the IProver interface. Good that we don't use it anywhere yet.

File: src/plugins/passwordTree/common/passwordTreeProgram.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+export const ProvePasswordInTreeProgram = Experimental.ZkProgram({
+    publicInput: MerkleRoot,
+    publicOutput: ProvePasswordInTreeOutput,
+
+    methods: {
+        baseCase: {
+            privateInputs: [PasswordInTreeWitness],
+            method(publicInput: MerkleRoot, privateInput: PasswordInTreeWitness): ProvePasswordInTreeOutput {
+                privateInput.witness
+                    .calculateRoot(Poseidon.hash([publicInput.root]))
```
Comment (adamczykm) [inline src/plugins/passwordTree/common/passwordTreeProgram.ts]: MISTAKE: use preimage

File: src/plugins/passwordTree/common/passwordTreeProgram.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+        inductiveCase: {
+            privateInputs: [SelfProof, PasswordInTreeWitness],
+            method(publicInput: MerkleRoot, earlierProof: SelfProof<MerkleRoot, ProvePasswordInTreeOutput>, privateInput: PasswordInTreeWitness): ProvePasswordInTreeOutput {
+                earlierProof.verify();
+                privateInput.witness
+                    .calculateRoot(Poseidon.hash([publicInput.root]))
+                    .assertEquals(publicInput.root);
+                return new ProvePasswordInTreeOutput(
+                    { recursiveMekleRootHash: Poseidon.hash([publicInput.root, earlierProof.publicOutput.recursiveMekleRootHash]) });
```
Comment (adamczykm) [inline src/plugins/passwordTree/common/passwordTreeProgram.ts]: rename to recursive hash

File: src/plugins/passwordTree/common/passwordTreeProgram.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                    { recursiveMekleRootHash: publicInput.root });
+            }
+        },
+
+        inductiveCase: {
+            privateInputs: [SelfProof, PasswordInTreeWitness],
+            method(publicInput: MerkleRoot, earlierProof: SelfProof<MerkleRoot, ProvePasswordInTreeOutput>, privateInput: PasswordInTreeWitness): ProvePasswordInTreeOutput {
+                earlierProof.verify();
+                privateInput.witness
+                    .calculateRoot(Poseidon.hash([publicInput.root]))
```
Comment (adamczykm) [inline src/plugins/passwordTree/common/passwordTreeProgram.ts]: Mistake: use preimage instead

File: src/plugins/passwordTree/common/passwordTreeProgram.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+export class ProvePasswordInTreeOutput extends Struct({
+    recursiveMekleRootHash: Field,
+}) {};
+
+// Prove knowledge of a preimage of a hash in a merkle tree.
+// The proof does not reveal the preimage nor the hash.
+// The output contains a recursive hash of all the roots for which the preimage is known.
+// output = hash(lastRoot + hash(secondLastRoot, ... hash(xLastRoot, lastRoot) ...)
+// Therefore the order of the proofs matters.
+export const ProvePasswordInTreeProgram = Experimental.ZkProgram({
```
Comment (adamczykm) [inline src/plugins/passwordTree/common/passwordTreeProgram.ts]: rename to e.g MerkleMemberships program

File: src/plugins/passwordTree/common/passwordTreeProgram.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import { Experimental, Field, MerkleWitness, Poseidon, SelfProof, Struct } from "o1js";
+
+// TODO how can this be made dynamic
+export const PASSWORD_TREE_HEIGHT = 10;
+
+export class PasswordTreeWitness extends MerkleWitness(PASSWORD_TREE_HEIGHT) {}
+
+export class PasswordInTreeWitness extends Struct({
+    witness: PasswordTreeWitness,
+    preImage: Field
```
Comment (adamczykm) [inline src/plugins/passwordTree/common/passwordTreeProgram.ts]: rename to "secret"

File: src/plugins/merkleMemberships/client/index.ts
Comment (adamczykm) [inline src/plugins/merkleMemberships/client/index.ts:1]: It's a nice functional implementation but got noisy. Typescript syntax is not very beautiful it appears. I'll propose some changes.

File: src/plugins/merkleMemberships/common/merkleMembershipsProgram.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
+import { Experimental, Field, MerkleWitness, Poseidon, SelfProof, Struct } from "o1js";
+
+// TODO how can this be made dynamic
+export const MERKLE_MEMBERSHIP_TREE_HEIGHT = 10;
+
+export class MerkleMembershipTreeWitness extends
```
Comment (adamczykm) [inline src/plugins/merkleMemberships/common/merkleMembershipsProgram.ts]: These are too wordy...

File: src/plugins/merkleMemberships/server/index.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+import { IMinAuthPlugin, IMinAuthPluginFactory } from "../../../library/plugin/pluginType";
+import { MinaTreesProvider, MinaTreesProviderConfiguration, TreesProvider, minaTreesProviderConfigurationSchema } from "./treeStorage";
+import { RequestHandler } from "express";
+import * as A from 'fp-ts/Array';
+
+const PoseidonHashSchema = z.bigint();
+
+const publicInputArgsSchema = z.array(PoseidonHashSchema);
+
+export type MerkleMembershipsPublicInputArgs =
```
Comment (adamczykm) [inline src/plugins/merkleMemberships/server/index.ts]: Wordiness again. Maybe we can drop the "MarkleMembership" prefix.

File: src/plugins/merkleMemberships/server/index.ts [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        root: root.toJSON(), 
+        witness: witness.toJSON(),
+      });
+
+    }
+  }
+
+  readonly publicInputArgsSchema = publicInputArgsSchema;
+
+  async verifyAndGetOutput(
```
Comment (adamczykm) [inline src/plugins/merkleMemberships/server/index.ts:81]: The proof is not being verified.
Please add the proof verification and code docstring that would explain what is going on.

Comment (adamczykm) [reply src/plugins/merkleMemberships/server/index.ts:81]: Rename `publicInputArgs` to `merkleRoots`.
Also, the argument is not an index of a tree it is its root.
The trees should be stored by the root hash and not the index.

File: tsconfig.json [hunk 13]
```diff
@@ -18,10 +18,6 @@
     "sourceMap": true,
     "noFallthroughCasesInSwitch": true,
     "allowSyntheticDefaultImports": true,
-    "baseUrl": ".",
-    "paths": {
-      "*": [ "src/*", "src/library/*" ]
```
Comment (adamczykm) [inline tsconfig.json:23]: let's add an alias for the library directory so that instead of ../.../../../../[path] we can write @lib/[path]

**Review by adamczykm — CHANGES_REQUESTED** [review]
There are some changes required.
Please try to add some docstrings already. They don't have to be beautiful yet, but should explain the high-level operations of the code. To make it easier to understand how the proof is achieved etc. 
Additionaly, we need the headless client to be able to test the flow of the updated plugin. 
It can be done with another PR though.

File: src/plugins/merkleMemberships/server/index.ts (hunk 12 above)
Comment (chfanghr) [reply src/plugins/merkleMemberships/server/index.ts:81]: > The proof is not being verified.

How so? The proof has already been verified by the server before entering this function. See https://github.com/mlabs-haskell/MinAuth/pull/7/files#diff-0b5d80d7710738d6c4dc8de490d335c59ba29ade54dd75b2027b3b3513c7452dR35-R69

File: src/library/tools/pluginServer/index.ts [hunk 14]
```diff
@@ -1,65 +1,73 @@
...
     // Use the appropriate plugin to verify the proof and return the output.
-    async function verifyProof(data: VerifyProofData): Promise<any> {
+    async function verifyProof(data: VerifyProofData): Promise<unknown> {
       const pluginName = data.plugin;
       console.info(`verifying proof using plugin ${pluginName}`);
       const pluginInstance = activePlugins[pluginName];
-      if (!pluginInstance)
-        throw `plugin ${pluginName} not found`;
+      if (!pluginInstance) throw `plugin ${pluginName} not found`;
       // Step 1: check that the proof was generated using a certain verification key.
```
Comment (adamczykm) [inline src/library/tools/pluginServer/index.ts]: This does not only verify that the proof was generated using the verification key and even if it does the gist is that it verifies that the proof is valid - it does not exit on broken assertions,  against some compiled source program. 
So maybe let's say:
```
// Step 1: check that the proof is valid against the given verification key.
```

File: src/plugins/merkleMemberships/client/index.ts [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+  private readonly cfg: ProverConfiguration;
+
+  async prove(
+    publicInput: Array<[ZkProgram.PublicInput, ZkProgram.TreeWitness]>,
+    secretInput: Array<Field>
+  ): Promise<JsonProof> {
+    if (publicInput.length != secretInput.length)
+      throw 'unmatched public/secret input list';
+
+    const proof: O.Option<ZkProof> = await A.reduce(
```
Comment (adamczykm) [inline src/plugins/merkleMemberships/client/index.ts:38]: Add a comment here please. Something akin to:
```
// For each pair of inputs (secret and public) add another layer of the recursive proof

File: src/plugins/merkleMemberships/client/index.ts [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+    return O.match(
+      () => {
+        throw 'empty input list';
+      }, // TODO: make it pure
+      (p: ZkProof) => p.toJSON()
+    )(proof);
+  }
+
+  async fetchPublicInputs(
```
Comment (adamczykm) [inline src/plugins/merkleMemberships/client/index.ts:100]: The problem here is that asking for a specific index of the tree reveals the identity.
The only solution is to transfer the entire tree. Then it should be searched for the index and this way the witness could be generated.
If there's no easy way to serialize the tree then the list of all leaves should be send and the tree should be rebuilt.

File: src/plugins/merkleMemberships/server/index.ts [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+export type PublicInputArgs = z.infer<typeof publicInputArgsSchema>;
+
+export class MerkleMembershipsPlugin
+  implements IMinAuthPlugin<PublicInputArgs, Field>
+{
+  readonly verificationKey: string;
+  private readonly storageProvider: TreesProvider;
+
+  customRoutes: Record<string, RequestHandler> = {
+    '/getWitness/:treeRoot/:leafIndex': async (req, resp) => {
```
Comment (adamczykm) [inline src/plugins/merkleMemberships/server/index.ts]: ditto:
the leafindex can be passsed to the server.

**Review by adamczykm — CHANGES_REQUESTED** [review]
The leaf index problem is still unresolved. The server should be sending the entire tree or all the leaves to ensure anonimity.

File: src/plugins/merkleMemberships/client/index.ts (hunk 16 above)
Comment (chfanghr) [reply src/plugins/merkleMemberships/client/index.ts:100]: This is apparently not very efficient. I don't see why this hurts anonymity - a client can always make a bunch of junk requests to confuse the server if the server decides to log the traffic. 

~~Aside from that, I have some concerns regarding the security of this implementation - if someone knows the value x where `Poseidon.hash([x]) == Field(0)` and if any of the trees have one or more holes, we are screwed.~~ Is this possible? Not so familiar with the poseidon hash algorithm.

**Review by adamczykm — APPROVED** [review]
Approving before the deeper review.
